feature_engine/selection/target_mean_encoding.py

```diff
@@ -0,0 +1,400 @@
+            2) Determine the mean value of the target within each label of the categorical variable using the train set
+            3) Use that mean target value per label as the prediction (using the test set) and calculate the roc-auc.
+
+        For each numerical variable:
+            1) Separate into train and test
+            2) Divide the variable into 100 quantiles
+            3) Calculate the mean target within each quantile using the training set 
+            4) Use that mean target value / bin as the prediction (using the test set) and calculate the roc-auc
+        
+        Implementation
+        --------------
+
+            Private methods
+            ---------------
+                `__target_mean_encoding(self, df_train_, df_test_, var_cols=None)`
+                `__validate_variables(self, X)`
+                    
+            Public methods
+            --------------
+                `fit(self, X, y=None)`
+                `transform(self)`
+                `fit_transform(self, X, y=None)`
+        
+    Parameters
+    ----------
+
+    variables: list, default=None
+        The list of variables to evaluate. If None, the transformer will evaluate all 
+        variables in the dataset associated with the variables_type.
+
+    target: string, default=None
+        The target variable to evaluate. If None, the transformer will return the 
+        relevant message for the missing value.
+
+    variables_type: boolean, default=True
+        If variables_type = True: the transformer will evaluate the variables as the 
+            categorical variables. 
+        If variables_type = False: the transformer will evaluate the variables as the
+            numerical variables.
+
+    metrics_score: string, default='roc_auc_score'
+        This indicates the metrics score to perform the feature selection.
+        The current support includes 'roc_auc_score' and 'r2_score'.
+
+    test_size: float, default=0.3
+        The test size setting of the data in the train_test_split method.
+
+    random_state: int, default=0
+        The random state setting in the train_test_split method.
+
+    quantiles: int, default=5
+        The amount of quantiles setting in the qcut of the fit method in case of the numerical types.
+        The range of the quantiles values is in between 3 and 100. The transformer will return a message
+        if the input quantiles value is out of this range.
+
+    """
+
+    def __init__(self, variables=None, target=None, variables_type=True, metrics_score='roc_auc_score', test_size=0.3, random_state=0, quantiles=5):
+        # inputs
+        self.variables = _define_variables(variables)
+        self.vars_type = variables_type
+        self.metrics_score = metrics_score
+        self.target = target
+        self.test_size = test_size
+        self.random_state = random_state
+        self.quantiles = quantiles
+
+        # transformed variables
+        self.target_enc_dict = {}
```

Comment:
these attributes should be initialised in the fit method, as they will be populated there and be followed by a trailing underscore, so that sklearn can recognise that the trasnformer has been fit.

Sklearn looks for attributes with trailing underscores to check that transformers were fit.

Also, I don't think we need to store the transformer columns. And I also think we should capitalise on feature-engine discretisers to perform the discretisation, instead of adding code manually. We have that functionality already.

Reply:
Can you please specify that should we use the single trailing underscore (a_) or double trailing underscore (__a__)? Sorry for this confusion. Thanks Sole!

Reply:
single trailing underscore `a_`